libpf/pfelf/file.go

```diff
@@ -418,20 +435,56 @@ func (f *File) Section(name string) *Section {
 	return nil
 }
 
+// findVirtualAddressProg determines the Prog header containing the virtual address.
+func (f *File) findVirtualAddressProg(addr uint64) *Prog {
+	// Search for the Program header that contains the start address.
+	for _, ph := range f.loadData {
+		if addr >= ph.Vaddr && addr < ph.Vaddr+ph.Memsz {
+			return ph
+		}
+	}
+	return nil
+}
+
+// VirtualMemory returns a slice for the request data at a virtual address.
+// The slice may point to mmapped data or be a newly allocated slice.
+func (f *File) VirtualMemory(addr int64, sz int) ([]byte, error) {
+	if sz == 0 {
+		return nil, nil
+	}
+	if ph := f.findVirtualAddressProg(uint64(addr)); ph != nil {
+		offset := addr - int64(ph.Vaddr)
+		if offset+int64(sz) <= int64(ph.Filesz) {
+			if mapping, ok := ph.elfReader.(*mmap.ReaderAt); ok {
+				return mapping.Subslice(int(ph.Off)+int(offset), sz)
+			}
+		}
+		buf := make([]byte, sz)
+		n, err := ph.ReadAt(buf, offset)
+		return buf[:n], err
+	}
+	return nil, fmt.Errorf("no matching segment for 0x%x", uint64(addr))
+}
+
+// SymbolData returns the data associated with given dynamic symbol.
+// The backing mmapped data is returned if possible, otherwise a maximum of
+// maxCopy bytes of the symbol data will read to newly allocated buffer.
+func (f *File) SymbolData(name libpf.SymbolName, maxCopy int) (libpf.SymbolValue, []byte, error) {
```

Comment:
`maxCopy` isn't used at all in this method so `VirtualMemory` will always allocate `sym.Size` if it creates a new slice. If we don't need to limit that (e.g. `maxCopy` < `sym.Size`), then we should remove this parameter and update all the callers. Otherwise, `VirtualMemory` needs to accept an extra parameter.